mindmeld/components/question_answerer.py

```diff
@@ -17,110 +17,148 @@
 import copy
 import json
 import logging
+import numbers
+import os
+import pickle
 import re
+import unicodedata
+import uuid
+import warnings
 from abc import ABC, abstractmethod
+from datetime import datetime
+from math import sin, cos, sqrt, atan2, radians
+from typing import List, Union
 
-from elasticsearch import ConnectionError as EsConnectionError
-from elasticsearch import ElasticsearchException, TransportError
+import nltk
+import numpy as np
+from nltk.corpus import stopwords as nltk_stopwords
+from nltk.stem import PorterStemmer
+from nltk.tokenize import word_tokenize as nltk_word_tokenize
 
-from ..exceptions import (
-    KnowledgeBaseConnectionError,
-    KnowledgeBaseError,
-    ElasticsearchVersionError,
-)
-from ..resource_loader import ResourceLoader
 from ._config import (
-    DEFAULT_ES_QA_MAPPING,
-    DEFAULT_RANKING_CONFIG,
     get_app_namespace,
     get_classifier_config,
 )
-from ._elasticsearch_helpers import (
-    DOC_TYPE,
-    create_es_client,
-    delete_index,
-    does_index_exist,
-    get_scoped_index_name,
-    load_index,
-    create_index_mapping,
-    is_es_version_7,
-    resolve_es_config_for_version,
+from ._util import _is_module_available, _get_module_or_attr as _getattr
+from .entity_resolver import EmbedderCosSimEntityResolver, TfIdfSparseCosSimEntityResolver
+from ..core import Bunch
+from ..exceptions import (
+    ElasticsearchKnowledgeBaseConnectionError,
+    KnowledgeBaseError,
+    ElasticsearchVersionError,
 )
 from ..models import create_embedder_model
+from ..path import (
+    get_question_answerer_index_cache_file_path,
+    NATIVE_QUESTION_ANSWERER_INDICES_CACHE_PATH
+)
+from ..resource_loader import Hasher, ResourceLoader
+
+if _is_module_available("elasticsearch"):
+    from ._elasticsearch_helpers import (
+        DOC_TYPE,
+        DEFAULT_ES_QA_MAPPING,
+        DEFAULT_ES_RANKING_CONFIG,
+        create_es_client,
+        delete_index,
+        does_index_exist,
+        get_scoped_index_name,
+        load_index,
+        create_index_mapping,
+        is_es_version_7,
+        resolve_es_config_for_version,
+    )
 
 logger = logging.getLogger(__name__)
 
-
 DEFAULT_QUERY_TYPE = "keyword"
-ALL_QUERY_TYPES = ["keyword", "text", "embedder", "embedder_keyword", "embedder_text"]
+ALL_QUERY_TYPES = ["keyword", "text", "embedder",
+                   "embedder_keyword", "keyword_embedder",
```

Comment:
do we need both `embedder_keyword` and `keyword_embedder`? (same for `embedder_text`)